### PR TITLE
plugins/refactoring: switch to mkNeovimPlugin & add telescope support option

### LINKED ISSUE #1218: feature: option to enable refactor.nvim's telescope extension
[refactoring.nvim](https://github.com/ThePrimeagen/refactoring.nvim) includes a Telescope extension that can be enabled using

```nix
extraConfigLuaPre = /* lua */ ''
  -- load refactoring Telescope extension
  require("telescope").load_extension("refactoring")
'';
```

And used in keymaps as follows:

```nix
keymps = [
  {
    mode = "n";
    key = "<leader>rr";
    action = /* lua */ ''
      function() require("telescope").extensions.refactoring.refactors() end
    '';
    lua = true;
    options.desc = "Select refactoring option";
  }
];
```

It'd be great if `plugins.refactoring` had options to configure this

#### Related

Perhaps `plugins.telescope.enabledExtensions` could be made public or an `plugins.telescope.extraExtensions` could be introduced to handle other unsupported telescope extensions?

### PR DESCRIPTION
**Pull request by MattSturgeon**
- [x] Switch to `mkNeovimPlugin`
- [x] Check settings options are up-to-date
- [x] Update tests
- [x] Add option to enable telescope extension 
- [x] Update option descriptions in line with current style and upstream readme

Fixes #1218

### CODE REVIEW

File: plugins/utils/refactoring.nix [hunk 1]
```diff
@@ -6,140 +6,163 @@
...
+      }
+    ];
 
-    promptFuncReturnType =
+    plugins.telescope.enabledExtensions = mkIf cfg.enableTelescope [ "refactoring" ];
+  };
+
+  settingsOptions = {
+    prompt_func_return_type =
       helpers.defaultNullOpts.mkNullable (with types; attrsOf bool)
```
Comment (traxys) [inline plugins/utils/refactoring.nix:45]: We should use mkAttrsOf if possible (this enables users to user raw lua)

File: plugins/utils/refactoring.nix [hunk 2]
```diff
@@ -6,140 +6,163 @@
...
+              cpp = true;
+              c = true;
+              java = true;
+            }
+          ```
         '';
 
-    promptFuncParamType =
+    prompt_func_param_type =
       helpers.defaultNullOpts.mkNullable (with types; attrsOf bool)
```
Comment (traxys) [inline plugins/utils/refactoring.nix:76]: mkAttrsOf

File: plugins/utils/refactoring.nix [hunk 3]
```diff
@@ -6,140 +6,163 @@
...
-            "c" = [];
-            "h" = [];
-            "hpp" = [];
-            "cxx" = [];
-          }
-        ''
-        ''
-          In any custom printf statement, it is possible to optionally add a max of one %s pattern, which is where the debug path will go.
-          For an example custom printf statement, go to this folder, select your language, and click on multiple-statements/printf.config.
+    printf_statements = helpers.defaultNullOpts.mkNullable (with types; attrsOf (listOf str)) "{ }" ''
```
Comment (traxys) [inline plugins/utils/refactoring.nix:105]: mkAttrsOf (listOf (maybeRaw str))

File: plugins/utils/refactoring.nix [hunk 4]
```diff
@@ -6,140 +6,163 @@
...
+          # add a custom printf statement for cpp
+          cpp = [ "std::cout << \"%s\" << std::endl;" ];
+        }
+      ```
+
+      [folder]: https://github.com/ThePrimeagen/refactoring.nvim/blob/master/lua/refactoring/tests/debug/printf
+    '';
+
+    print_var_statements =
+      helpers.defaultNullOpts.mkNullable (with types; attrsOf (listOf str)) "{ }"
```
Comment (traxys) [inline plugins/utils/refactoring.nix:125]: mkAttrsOf (listOf (maybeRaw str))

File: plugins/utils/refactoring.nix [hunk 5]
```diff
@@ -6,140 +6,163 @@
...
+              # add a custom print var statement for cpp
+              cpp = [ "printf(\"a custom statement %%s %s\", %s)" ];
+            }
+          ```
+
+          [folder]: https://github.com/ThePrimeagen/refactoring.nvim/blob/master/lua/refactoring/tests/debug/print_var
         '';
-    };
+
+    extract_var_statements = helpers.defaultNullOpts.mkNullable (with types; attrsOf str) "{ }" ''
```
Comment (traxys) [inline plugins/utils/refactoring.nix:146]: mkAttrsOf str

**Review by traxys — CHANGES_REQUESTED** [review]

File: plugins/utils/refactoring.nix (hunk 1 above)
Comment (MattSturgeon) [reply plugins/utils/refactoring.nix]: :+1:

**Review by traxys — APPROVED** [review]

**Pull request merged**